packages/remirror__extension-file/src/file-placeholder-plugin.ts

```diff
@@ -0,0 +1,71 @@
+import { EditorState, Plugin, Transaction } from '@remirror/pm/state';
+import { Decoration, DecorationSet } from '@remirror/pm/view';
+
+interface PlaceholderPluginAction {
+  add?: {
+    id: unknown;
+    payload: any;
+    pos: number;
+  };
+  remove?: {
+    id: unknown;
+  };
+}
+
+interface PlaceholderPluginData {
+  set: DecorationSet;
+  payloads: Map<unknown, any>;
+}
+
+export const placeholderPlugin = new Plugin<PlaceholderPluginData>({
+  state: {
+    init(): PlaceholderPluginData {
+      return { set: DecorationSet.empty, payloads: new Map<unknown, any>() };
+    },
+    apply(tr, { set, payloads }: PlaceholderPluginData): PlaceholderPluginData {
+      // Adjust decoration positions to changes made by the transaction
+      set = set.map(tr.mapping, tr.doc);
+      // See if the transaction adds or removes any placeholders
+      const action = tr.getMeta(this) as PlaceholderPluginAction | null;
+
+      if (action) {
+        const { add, remove } = action;
```

Comment:
For consistency, I think we should use the same pattern as the [Annotation plugin](https://github.com/remirror/remirror/blob/beta/packages/remirror__extension-annotation/src/annotation-plugin.ts#L120-L142).

I.e. Instead of `action.add` or `action.remove` we have `action.type = ADD_PLACEHOLDER`/`action.type = REMOVE_PLACEHOLDER`

```ts
if (action.type === ADD_PLACEHOLDER) {
  // Do stuff
}

if (action.type === REMOVE_PLACEHOLDER) {
  // Do stuff
}
```

Reply:
done